src/utils/convert3Dto4D.m

```diff
@@ -0,0 +1,200 @@
+
+  % 0:  SAME
+  % 2:  UINT8   - unsigned char
+  % 4:  INT16   - signed short
+  % 8:  INT32   - signed int
+  % 16: FLOAT32 - single prec. float
+  % 64: FLOAT64 - double prec. float
+
+  optSource.dataType = 0;
+
+  % Boolean to enable gzip of the new 4D file (0 is default)
+  optSource.zip = 0;
+
+  % Check the options provided
+  optSource = checkOptionsSource(optSource);
+
+  % Get source folder content
+  sourceDataStruc = dir(optSource.sourceDir);
+
+  isDir = [sourceDataStruc(:).isdir];
+
+  optSource.sequenceList = {sourceDataStruc(isDir).name}';
+
+  % Loop through the sequence folders
+
+  tic;
+
+  for iSeq = 1:size(optSource.sequenceList, 1)
+
+    % Skip 'non' folders
+    if length(char(optSource.sequenceList(iSeq))) > 2
+
+      % Check if sequence to ignore or not
+      if contains(optSource.sequenceList(iSeq), optSource.sequenceToIgnore)
+
+        newline;
+        warning('IGNORING SEQUENCE: %s', string(optSource.sequenceList(iSeq)));
+        newline;
+
+      else
+
+        fprintf('\n\nCONVERTING SEQUENCE: %s \n\n', char(optSource.sequenceList(iSeq)));
+
+        % Set whether to remove dummies or not
+
+        nbDummies = 0;
+
+        if contains(optSource.sequenceList(iSeq), optSource.sequenceRmDummies)
+
+          nbDummies = optSource.nbDummies;
+
+          fprintf('\n\nREMOVING %s DUMMIES\n\n', num2str(nbDummies));
+
+        end
+
+        % Get sequence folder path
+        sequencePath = char(fullfile(optSource.sourceDir, optSource.sequenceList(iSeq)));
+
+        % Retrieve volume files info
+        [volumesList, outputNameImage] = parseFiles('nii', sequencePath, nbDummies);
+
+        % Set output name, it takes the file name of the 1st volume of the 4D file and add subfix
+        outputNameImage = strrep(outputNameImage, '.nii', '_4D.nii');
+
+        % Retrieve sidecar json files info
+        [jsonList, outputNameJson] = parseFiles('json', sequencePath, nbDummies);
+
+        jsonFile = spm_jsonread(char(jsonList(1)));
+
+        % % % % % % LIEGE  SPECIFIC % % % % % % %
+        RT = jsonFile.acqpar.RepetitionTime / 1000;
+        % % % % % % % % % % % % % % % % % % % % %
+
+        % Set and run spm batch, input all the volumes minus the dummies if > 0
+        matlabbatch = setBatch3Dto4D(volumesList(nbDummies + 1:end, :), ...
+                                     outputNameImage, ...
+                                     optSource.dataType, ...
+                                     RT);
+
+        spm_jobman('run', matlabbatch);
+
+        if optSource.zip
+
+          % Zip and delete the and the new 4D file
+          fprintf(1, 'ZIP AND DELETE THE NEW 4D BRAIN \n\n');
+
+          gzip([sequencePath filesep outputNameImage]);
+
+          delete([sequencePath filesep outputNameImage]);
+
+        end
+
+        % Save one sidecar json file, it takes the file name of the 1st volume of the 4D file and
+        % add subfix
+        if ~isempty(jsonList)
+
```

Comment:
```suggestion
% TODO: cover the MoCo use case
% - if the sequence is MoCo (motion corrected when the "scanner" reconstructs the images - an option on can tick
% on Siemens scanner and that output an additional MoCo file with the regular sequence) then each JSON file of 
% each volume contains the motion correction information for that volume.
% So only taking the JSON of the first volume means we "lose" the realignment parameters that could be useful later.
```

Just adding a TODO here for our future selves: this is more of an "edge" case as of now I am not even sure that BIDS supports having the motion correction information in the JSON (but I will open an issue on the specs repo).

Reply:
ok! very good to know